lib/timers.js

```diff
@@ -440,13 +452,20 @@ function rearm(timer) {
 
 
 const clearTimeout = exports.clearTimeout = function(timer) {
-  if (timer && (timer[kOnTimeout] || timer._onTimeout)) {
+  if (!timer) return;
+
+  if (timer[kOnTimeout] || timer._onTimeout) {
+    if (typeof timer._onTimeout !== 'function') {
+      promiseResolve(timer._onTimeout, undefined);
+    }
     timer[kOnTimeout] = timer._onTimeout = null;
     if (timer instanceof Timeout) {
       timer.close(); // for after === 0
     } else {
       unenroll(timer);
     }
+  } else if (isPromise(timer)) {
```

Comment:
First of all, attaching properties to native objects seems like a bad idea from performance perspective. Secondly, this implements ad hoc cancellation mechanism, while there is no spec on standards track or even community consensus. This approach suffers from the fact that it's not propagated and there it doesn't play nice with async-await.

I think it's fine to make promisified versions not cancellable. If you use them you are most likely not going to cancel anyway.